### ISSUE TITLE
timelocal not -w safe

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7776]
Migrated from [rt.perl.org#33989](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=33989) (status was 'resolved')

Searchable as RT33989$

**Comment by p5pRT** [comment]
### From sune.karlsson@esi.oru.se

#### Created by sune.karlsson@esi.oru.se



timelocal in version 1\.10 of Time&#8203;::Local is not \-w safe&#8203;:

c&#8203;:\\temp>perl \-MTime&#8203;::Local \-we "print timelocal\(localtime\(\)\)"
Use of uninitialized value in integer addition \(\+\) at C&#8203;:/Program/Perl/lib/Time/Local\.pm line 76\.
Use of uninitialized value in integer multiplication \(\*\) at C&#8203;:/Program/Perl/lib/Time/Local\.pm line 76\.
Use of uninitialized value in integer multiplication \(\*\) at C&#8203;:/Program/Perl/lib/Time/Local\.pm line 76\.
Use of uninitialized value in pack at C&#8203;:/Program/Perl/lib/Time/Local\.pm line 67\.

Use of uninitialized value in pack at C&#8203;:/Program/Perl/lib/Time/Local\.pm line 67\.

Use of uninitialized value in integer addition \(\+\) at C&#8203;:/Program/Perl/lib/Time/Local\.pm line 68\.
Use of uninitialized value in integer addition \(\+\) at C&#8203;:/Program/Perl/lib/Time/Local\.pm line 69\.
Use of uninitialized value in integer addition \(\+\) at C&#8203;:/Program/Perl/lib/Time/Local\.pm line 67\.
1107103782

As far as I can tell this comes from the second call to \_zoneadjust \(line 77\) where
$MinInt \(= \-2147483648\) is fed to localtime which doesn't work to well with negative arguments\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=medium

Site configuration information for perl v5.8.6:

Configured by ActiveState at Mon Dec 13 09:51:32 2004.

Summary of my perl5 (revision 5 version 8 subversion 6) configuration:
  Platform:
    osname=MSWin32, osvers=4.0, archname=MSWin32-x86-multi-thread
    uname=''
    config_args='undef'
    hint=recommended, useposix=true, d_sigaction=undef
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cl', ccflags ='-nologo -Gf -W3 -MD -Zi -DNDEBUG -O1 -DWIN32 -D_CONSOLE -DNO_STRICT -DHAVE_DES_FCRYPT  -DNO_HASH_SEED -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -DPERL_MSVCRT_READFIX',
    optimize='-MD -Zi -DNDEBUG -O1',
    cppflags='-DWIN32'
    ccversion='', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=10
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='__int64', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='link', ldflags ='-nologo -nodefaultlib -debug -opt:ref,icf  -libpath:"c:\Program\Perl\lib\CORE"  -machine:x86'
    libpth=\lib
    libs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib msvcrt.lib
    perllibs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib msvcrt.lib
    libc=msvcrt.lib, so=dll, useshrplib=yes, libperl=perl58.lib
    gnulibc_version='undef'
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -debug -opt:ref,icf  -libpath:"c:\Program\Perl\lib\CORE"  -machine:x86'

Locally applied patches:
    ACTIVEPERL_LOCAL_PATCHES_ENTRY
    21540 Fix backward-compatibility issues in if.pm
    23565 Wrong MANIFEST.SKIP


@INC for perl v5.8.6:
    c:\MyDocs\code\perl
    C:/Program/Perl/lib
    C:/Program/Perl/site/lib
    .


Environment for perl v5.8.6:
    HOME (unset)
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=C:\Program\Perl\bin\;c:\MyDocs\code\perl;C:\Program Files\Microsoft Visual Studio\Common\Tools;C:\Program Files\Microsoft Visual Studio\Common\Msdev98\BIN;C:\Program Files\Microsoft Visual Studio\DF98\BIN;C:\Program Files\Microsoft Visual Studio\VC98\BIN;C:\program\miktex\texmf\miktex\bin;C:\Program\Windows Resource Kits\Tools\;C:\PROGRAM\THINKPAD\UTILITIES;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\Program\ATI Technologies\ATI Control Panel;C:\Program\PC-Doctor for Windows\services;C:\Program\Support Tools\;C:\Program\ActiveState Komodo 2.5\
    PERL5LIB=c:\MyDocs\code\perl
    PERL_BADLANG (unset)
    SHELL (unset)





```
</details>

**Comment by p5pRT** [comment]
### From @steve-m-hay

sune\.karlsson@&#8203;esi\.oru\.se \(via RT\) wrote&#8203;:

>\# New Ticket Created by  sune\.karlsson@&#8203;esi\.oru\.se 
>\# Please include the string&#8203;:  \[perl \#33989\]
>\# in the subject line of all future correspondence about this issue\. 
>\# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=33989 >
>
>
>This is a bug report for perl from sune\.karlsson@&#8203;esi\.oru\.se\,
>generated with the help of perlbug 1\.35 running under perl v5\.8\.6\.
>
>
>\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>\[Please enter your report here\]
>
>timelocal in version 1\.10 of Time&#8203;::Local is not \-w safe&#8203;:
>
>c&#8203;:\\temp>perl \-MTime&#8203;::Local \-we "print timelocal\(localtime\(\)\)"
>Use of uninitialized value in integer addition \(\+\) at C&#8203;:/Program/Perl/lib/Time/Local\.pm line 76\.
>Use of uninitialized value in integer multiplication \(\*\) at C&#8203;:/Program/Perl/lib/Time/Local\.pm line 76\.
>Use of uninitialized value in integer multiplication \(\*\) at C&#8203;:/Program/Perl/lib/Time/Local\.pm line 76\.
>Use of uninitialized value in pack at C&#8203;:/Program/Perl/lib/Time/Local\.pm line 67\.
>
>Use of uninitialized value in pack at C&#8203;:/Program/Perl/lib/Time/Local\.pm line 67\.
>
>Use of uninitialized value in integer addition \(\+\) at C&#8203;:/Program/Perl/lib/Time/Local\.pm line 68\.
>Use of uninitialized value in integer addition \(\+\) at C&#8203;:/Program/Perl/lib/Time/Local\.pm line 69\.
>Use of uninitialized value in integer addition \(\+\) at C&#8203;:/Program/Perl/lib/Time/Local\.pm line 67\.
>1107103782
>
>As far as I can tell this comes from the second call to \_zoneadjust \(line 77\) where
>$MinInt \(= \-2147483648\) is fed to localtime which doesn't work to well with negative arguments\.
>
Indeed\. In fact\, looking at the MS CRT source code \(localtim\.c\) we see 
the following note&#8203;:

\*       \(3\) It is assumed that time\_t is a 32\-bit long integer representing
\*           the number of seconds since 00&#8203;:00&#8203;:00\, 01\-01\-70 \(UTC\) \(i\.e\.\, the
\*           Posix/Unix Epoch\. Only non\-negative values are supported\.

and sure enough\, Perl's localtime\(\) returns nothing when fed any 
negative values\.

The attached patch \(against Time\-Local\-1\.10\) fixes the example above\, 
and also silences a bucketload of similar noise that was previously 
coming from Time&#8203;::Local when running the LWP test suite on Win32\.

\- Steve




---
Radan Computational Ltd\.

The information contained in this message and any files transmitted with it are confidential and intended for the addressee\(s\) only\.  If you have received this message in error or there are any problems\, please notify the sender immediately\.  The unauthorized use\, disclosure\, copying or alteration of this message is strictly forbidden\.  Note that any views or opinions presented in this email are solely those of the author and do not necessarily represent those of Radan Computational Ltd\.  The recipient\(s\) of this message should check it and any attached files for viruses&#8203;: Radan Computational will accept no liability for any damage caused by any virus transmitted by this email\.

**Comment by p5pRT** [comment]
### From @steve-m-hay

<details><summary>Inline Patch</summary>

```diff
--- lib/Time/Local.pm.orig	2004-04-21 01:04:06.000000000 +0100
+++ lib/Time/Local.pm	2005-02-08 11:06:33.653485500 +0000
@@ -30,6 +30,10 @@
     # time_t is unsigned...
     $MaxInt = (1 << (8 * $Config{intsize})) - 1;
     $MinInt = 0;
+} elsif ($^O eq 'MSWin32') {
+    # time_t is signed but localtime() only supports non-negative input...
+    $MaxInt = ((1 << (8 * $Config{intsize} - 2))-1)*2 + 1;
+    $MinInt = 0;
 } else {
     $MaxInt = ((1 << (8 * $Config{intsize} - 2))-1)*2 + 1;
     $MinInt = -$MaxInt - 1;
```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @rgs

Steve Hay wrote&#8203;:
> The attached patch \(against Time\-Local\-1\.10\) fixes the example above\, 
> and also silences a bucketload of similar noise that was previously 
> coming from Time&#8203;::Local when running the LWP test suite on Win32\.

Added in Time\-Local rt\.cpan queue\, closing it here\.

**Comment by p5pRT** [comment]
### From @autarch

On Tue\, 8 Feb 2005\, Steve Hay wrote&#8203;:

> Indeed\. In fact\, looking at the MS CRT source code \(localtim\.c\) we see
> the following note&#8203;:
>
> \*       \(3\) It is assumed that time\_t is a 32\-bit long integer representing
> \*           the number of seconds since 00&#8203;:00&#8203;:00\, 01\-01\-70 \(UTC\) \(i\.e\.\, the
> \*           Posix/Unix Epoch\. Only non\-negative values are supported\.
>
> and sure enough\, Perl's localtime\(\) returns nothing when fed any
> negative values\.
>
> The attached patch \(against Time\-Local\-1\.10\) fixes the example above\,
> and also silences a bucketload of similar noise that was previously
> coming from Time&#8203;::Local when running the LWP test suite on Win32\.
>
> \- Steve

This is fixed in the CVS version of Time&#8203;::Local\.  See 
http&#8203;://sourceforge\.net/cvs/?group\_id=17187 for details\.

I was hoping to get some volunteer testers on VMS and other "exotic" 
platforms before a release\, but no one responded\, so I'll probably release 
1\.11 soon\.


\-dave

/\*===================================================
VegGuide\.Org                        www\.BookIRead\.com
Your guide to all that's veg\.       My book blog
===================================================\*/

**Comment by p5pRT** [comment]
### From @tux

On Tue\, 8 Feb 2005 11&#8203;:08&#8203;:22 \-0600 \(CST\)\, Dave Rolsky \<autarch@&#8203;urth\.org> wrote&#8203;:

> On Tue\, 8 Feb 2005\, Steve Hay wrote&#8203;:
> 
> > Indeed\. In fact\, looking at the MS CRT source code \(localtim\.c\) we see
> > the following note&#8203;:
> >
> > \*       \(3\) It is assumed that time\_t is a 32\-bit long integer
> > representing \*           the number of seconds since 00&#8203;:00&#8203;:00\, 01\-01\-70
> > \(UTC\) \(i\.e\.\, the \*           Posix/Unix Epoch\. Only non\-negative values
> > are supported\.
> >
> > and sure enough\, Perl's localtime\(\) returns nothing when fed any
> > negative values\.
> >
> > The attached patch \(against Time\-Local\-1\.10\) fixes the example above\,
> > and also silences a bucketload of similar noise that was previously
> > coming from Time&#8203;::Local when running the LWP test suite on Win32\.
> >
> > \- Steve
> 
> This is fixed in the CVS version of Time&#8203;::Local\.  See 
> http&#8203;://sourceforge\.net/cvs/?group\_id=17187 for details\.

Now that you mention Time&#8203;::Local\, I get this failure on AIX 5\.2 with vac
for blead&#8203;:

\-\-8\<\-\-\-
\.\./lib/Time/Local\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\# Test 102 got&#8203;: "3c19137f" \(\.\./lib/Time/Local\.t at line 108\)
\#     Expected&#8203;: "7fffffff" \(0x7fffffff round trip through localtime then timelocal\)
\#  \.\./lib/Time/Local\.t line 108 is&#8203;:     ok\(sprintf\('%x'\, timelocal\(localtime\(0x7fffffff\)\)\)\, sprintf\('%x'\, 0x7fffffff\)\,
\.\./lib/Time/Local\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.FAILED test 102
&nbsp;        Failed 1/104 tests\, 99\.04% okay
\-\->8\-\-\-

> I was hoping to get some volunteer testers on VMS and other "exotic" 
> platforms before a release\, but no one responded\, so I'll probably release 
> 1\.11 soon\.

\-\- 
H\.Merijn Brand        Amsterdam Perl Mongers \(http&#8203;://amsterdam\.pm\.org/\)
using Perl 5\.6\.2\, 5\.8\.0\, 5\.8\.3\, & 5\.9\.2  on HP\-UX 10\.20\, 11\.00 & 11\.11\,
&nbsp;  AIX 4\.3\, SuSE 9\.0 pro 2\.4\.21 & Win2k\.     http&#8203;://www\.cmve\.net/~merijn
Smoking perl&#8203;: smokers@&#8203;perl\.org\,             perl QA&#8203;: http&#8203;://qa\.perl\.org
&nbsp;  reports to&#8203;: smokers\-reports@&#8203;perl\.org\,                perl\-qa@&#8203;perl\.org
